Fix SETUP generating invalid files.

Diman0 opened the pull request:
## What type of PR?

bug-fix

## What does this PR do?
If JavaScript script is disabled or JS files are blocked from the setup website, the SETUP website would generate invalid configuration. Checks were added to prevent this situation from occurring.

Bootstrap loaded an outdated jquery version 1. We already use Jquery 2. I have disabled this and configured bootstrap to load our jquery js file.

Also fixed a regular expression that makes sure the specified Mailu storage path is valid.

### Related issue(s)
- closes #2396
- closes #2890
- closes #1524  

## Prerequisites
Before we can consider review and merge, please make sure the following list is done and checked.
If an entry in not applicable, you can check it or remove it from the list.

- [n/a] In case of feature or enhancement: documentation updated accordingly
- [x] Unless it's docs or a minor change: add [changelog](https://mailu.io/master/contributors/workflow.html#changelog) entry file.

nextgens commented:
Can we get #1524 nailed too? What about doing the check server side and erroring out if we don't get parameters instead?

Diman0 commented:
> Can we get #1524 nailed too? What about doing the check server side and erroring out if we don't get parameters instead?

I wasn't aware of that issue. Yes. That should be included as well.

Diman0 commented:
@nextgens I have updated the hint for setting the subnet and changed the regular expression so it only accepts a 0 for the fourth parameter. So this format is accepted *.*.*.0/** (where * is a number 0-9).

How would server side validation look like? I don't have a lot experience with web development. Does it offer advantages?
 I  don't want to spend too much time on this. In the future we want to get rid of setup after all.

nextgens commented:
> @nextgens I have updated the hint for setting the subnet and changed the regular expression so it only accepts a 0 for the fourth parameter. So this format is accepted _._.*.0/** (where * is a number 0-9).
> 

That may break things for those that do not use a /24 ... we could just switch strict to True here
 https://github.com/Mailu/Mailu/blob/master/setup/server.py#L94 ... and return a proper error.

> How would server side validation look like? I don't have a lot experience with web development. Does it offer advantages? I don't want to spend too much time on this. In the future we want to get rid of setup after all.

The main advantage would be that it would work in all scenarios (including when part of the JS is disabled because of some browser extension).

Well we would double-check if we do get the parameters we expect... and if we don't we flash() an error message.

Diman0 commented:
> > @nextgens I have updated the hint for setting the subnet and changed the regular expression so it only accepts a 0 for the fourth parameter. So this format is accepted _._.*.0/** (where * is a number 0-9).
> 
> That may break things for those that do not use a /24 ... we could just implement something like https://github.com/Mailu/Mailu/blob/master/setup/server.py#L94 ... and return a proper error.
> 

I'm probably missing something, what would exactly break? The reg exp only enforces a 0 for the 4th parameter. So you can use:

* 192.168.12.0/24
* 192.168.0.0/16
* 192.0.0.0/8

nextgens commented:
You can't use 192.168.0.128/25 for instance. Subnets may have less than 256 addresses and enforcing that there is a zero as the 4th parameter is technically the wrong test. What you want to do is ensure that the "host bits" are set to zero... but you can't do it easily using a regexp.

Mind you maybe this is advanced-enough usage that we don't mind it being broken.

Diman0 commented:
I didn't know that you could create such (limited) subnets. I think it is too advanced for Mailu. The suggested subnet should only be changed if it already exists. The main goal is to prevent less experienced users from making a mistake by picking 192.111.12.1/24 as subnet.

nextgens commented:
Another way to deal with this would be to read it server-side and output the right thing in the config (fix it for the user).

https://docs.python.org/3/library/ipaddress.html#ipaddress.ip_network

Diman0 commented:
Ah that function checks if the provided IP is a valid subnet. I did not realize that python offered this functionality. This makes it easy to check it server-side and show a human readable message that the subnet is invalid. I will adapt setup to do this.

Diman0 commented:
bors try

nextgens commented:
let's do the same for SUBNET6 below, no?

nextgens commented:
I'd get rid of the scary regexp at https://github.com/Mailu/Mailu/pull/3025/files#diff-11e701060cb3c1f0e208aa5ad066ac140e32551bcd8ab794d795669c4d40941bL38 too

Diman0 commented:
bors try

Diman0 commented:
discussed during meeting.
By default make div container hidden. The text you need javascript is shown by default. Via onloaded event use javascript (something from render.js maybe) to make the container div visible and the warning invisible. So if javascript is blocked you still see the warning.

Do all validation server side. So add checks for everything in `def submit()`

Diman0 commented:
bors try

nextgens commented on setup/Dockerfile:
```diff
@@ -14,6 +14,10 @@ COPY main.py ./main.py
 
+#Note: This is appended so we can explicitly check if this JS file has been loaded
+#by the user's internet browser when accessing the setup site.
+RUN echo var jQueryMailu=\'loaded\'\; >> ./static/jquery.min.js
```
Not required anymore

nextgens commented on setup/templates/steps/config.html:
```diff
@@ -58,13 +58,6 @@
-<div class="form-check form-check-inline">
-  <label class="form-check-label">
-    <input class="form-check-input" type="checkbox" name="disable_statistics" value="True">
-    Opt-out of statistics
```
Hmm, I wouldn't remove this without removing the code in admin. Can't we fix it instead?

Diman0 replied:
It has never worked. It requires a separate webserver to run (https://github.com/Mailu/Stats). This webserver has never been deployed.  

But I agree, if this is removed then the code in admin should be removed as well. I didn't think about admin.

Diman0 replied:
I could reinstate the code, but will simply hide the element and set it to disabled in mailu.env. 
That way we can still choose if we want to implement it in the future by hosting the webserver required for this.

nextgens commented on setup/flavors/compose/mailu.env:
```diff
@@ -39,9 +39,6 @@ AUTH_RATELIMIT_IP={{ auth_ratelimit_ip }}/hour
 {% endif %}
 
-# Opt-out of statistics, replace with "True" to opt out
-DISABLE_STATISTICS={{ disable_statistics or 'False' }}
```
Same thing, I wouldn't remove it if we are still making the request

Diman0 commented:
bors try

nextgens commented on setup/server.py:
```diff
@@ -90,12 +94,65 @@ def wizard():
         data['uid'] = str(uuid.uuid4())
+        valid = True
+        try:
+            ipaddress.ip_address(data['bind4'])
```
ipaddress.IPv4Address would be a better choice here

nextgens commented on setup/server.py:
```diff
@@ -90,12 +94,65 @@ def wizard():
+            flask.flash('Configured IPv4 address is invalid', 'error')
+            valid = False
+        try:
+            ipaddress.ip_address(data['bind6'])
```
same here and below for ip_network

nextgens commented on setup/static/render.js:
```diff
@@ -1,21 +1,30 @@
 	if ($("#webmail").val() == 'none') {
 		$("#webmail_path").hide();
 		$("#webmail_path").val("");
+		$("#webmail_path").prop('required',false);
```
I would rather get rid of all ``WEB_`` variables tbh. We agreed Mailu should have it's own FQDN, let's simplify and hard-code paths on it.

Diman0 replied:
I'm fine with removing it from setup only. Setup should be simplified and help the end-user not making any mistakes. A typical user should not have to change the `WEB_` settings. 

Advanced users can still tweak mailu.env.

nextgens commented on setup/templates/steps/compose/03_expose.html:
```diff
@@ -16,11 +16,10 @@
-    value="192.168.203.0/24">
+  <input class="form-control" type="text" name="bind4" value="127.0.0.1" >
+  <label>Subnet of the docker network. This should not conflict with any networks to which your system is connected. (Internal and external!)
+    Fourth number is always 0. Normally the format is '*.*.*.0/24'</label>
```
```suggestion
    Usually the format is '*.*.*.0/24'</label>
```

Diman0 commented:
bors try

Diman0 commented:
bors try-

Diman0 commented:
bors try

nextgens commented:
@Diman0 should we backport this?

Diman0 commented:
> @Diman0 should we backport this?

That is not a bad idea. The issue was reported for mailu 2.0 after all.

nextgens commented:
@mergifyio backport

nextgens commented:
@mergifyio backport 2.0
